cmd/kubeadm/app/phases/upgrade/health.go

```diff
@@ -59,49 +64,62 @@ func (c *healthCheck) Name() string {
 // - all control-plane Nodes are Ready
 // - (if self-hosted) that there are DaemonSets with at least one Pod for all control plane components
 // - (if static pod-hosted) that all required Static Pod manifests exist on disk
-func CheckClusterHealth(client clientset.Interface, ignoreChecksErrors sets.String) error {
-	fmt.Println("[upgrade] Making sure the cluster is healthy:")
+func CheckClusterHealth(client clientset.Interface, cfg *kubeadmapi.ClusterConfiguration, ignoreChecksErrors sets.String) error {
+	fmt.Println("[upgrade] Running cluster health checks")
 
 	healthChecks := []preflight.Checker{
 		&healthCheck{
-			name:   "APIServerHealth",
+			name:   "DeamonSetCreate",
 			client: client,
-			f:      apiServerHealthy,
+			cfg:    cfg,
+			f:      daemonSetPrepullCheck,
 		},
 		&healthCheck{
 			name:   "ControlPlaneNodesReady",
 			client: client,
 			f:      controlPlaneNodesReady,
 		},
-		// TODO: Add a check for ComponentStatuses here?
+		&healthCheck{
+			name:   "StaticPodManifest",
+			client: client,
+			cfg:    cfg,
+			f:      staticPodManifestHealth,
+		},
 	}
 
-	healthChecks = append(healthChecks, &healthCheck{
-		name:   "StaticPodManifest",
-		client: client,
-		f:      staticPodManifestHealth,
-	})
-
 	return preflight.RunChecks(healthChecks, os.Stderr, ignoreChecksErrors)
 }
 
-// apiServerHealthy checks whether the API server's /healthz endpoint is healthy
-func apiServerHealthy(client clientset.Interface) error {
-	healthStatus := 0
+// daemonSetPrepullCheck checks if a DaemonSet that uses the pause container can be created in the cluster.
+func daemonSetPrepullCheck(client clientset.Interface, cfg *kubeadmapi.ClusterConfiguration) error {
```

Comment:
This is not a pre-pull check. The only way it's going to resort to pulling something is if we bump the pause image version.

Reply:
updated.